File: ios/Photo/UI/ZoomableImageView.swift
import UIKit

/// UIImageView supporting pinch-to-zoom and single-finger pan for the photo
/// preview, reporting the current on-screen image rect via `onBoundsChanged`
/// so the crop overlay can stay aligned to it.
final class ZoomableImageView: UIImageView {
  var onBoundsChanged: ((CGRect) -> Void)?
  var panZoomEnabled = true {
    didSet {
      pinchRecognizer.isEnabled = panZoomEnabled
      panRecognizer.isEnabled = panZoomEnabled
    }
  }

  private let pinchRecognizer = UIPinchGestureRecognizer()
  private let panRecognizer = UIPanGestureRecognizer()
  private var minScale: CGFloat = 1
  private var maxScale: CGFloat = 4

  override init(frame: CGRect) {
    super.init(frame: frame)
    contentMode = .scaleAspectFit
    isUserInteractionEnabled = true
    pinchRecognizer.addTarget(self, action: #selector(handlePinch(_:)))
    panRecognizer.addTarget(self, action: #selector(handlePan(_:)))
    addGestureRecognizer(pinchRecognizer)
    addGestureRecognizer(panRecognizer)
  }

  required init?(coder: NSCoder) { nil }

  /// Fits the current image to the view bounds and resets zoom/pan.
  func resetToFit() {
    transform = .identity
    minScale = 1
    maxScale = 4
    reportBounds()
  }

  func currentImageBounds() -> CGRect {
    guard let image = image, bounds.width > 0, bounds.height > 0, image.size.width > 0, image.size.height > 0 else { return bounds }
    let scale = min(bounds.width / image.size.width, bounds.height / image.size.height)
    let fittedSize = CGSize(width: image.size.width * scale, height: image.size.height * scale)
    let origin = CGPoint(x: (bounds.width - fittedSize.width) / 2, y: (bounds.height - fittedSize.height) / 2)
    return CGRect(origin: origin, size: fittedSize).applying(transform)
  }

  private func reportBounds() {
    onBoundsChanged?(currentImageBounds())
  }

  private func currentScaleFactor() -> CGFloat {
    sqrt(transform.a * transform.a + transform.c * transform.c)
  }

  @objc private func handlePinch(_ gesture: UIPinchGestureRecognizer) {
    guard panZoomEnabled else { return }
    let currentScale = currentScaleFactor()
    let targetScale = max(minScale, min(maxScale, currentScale * gesture.scale))
    let appliedFactor = targetScale / currentScale
    let location = gesture.location(in: self)
    let anchor = CGPoint(x: location.x - bounds.midX, y: location.y - bounds.midY)
    transform = transform
      .translatedBy(x: anchor.x, y: anchor.y)
      .scaledBy(x: appliedFactor, y: appliedFactor)
      .translatedBy(x: -anchor.x, y: -anchor.y)
    gesture.scale = 1
    reportBounds()
  }

  @objc private func handlePan(_ gesture: UIPanGestureRecognizer) {
    guard panZoomEnabled else { return }
    let translation = gesture.translation(in: self)
    let scale = currentScaleFactor()
    transform = transform.translatedBy(x: translation.x / scale, y: translation.y / scale)
    gesture.setTranslation(.zero, in: self)
    reportBounds()
  }
}

File: android/src/main/java/com/photovideoeditor/photo/ui/ZoomableImageView.kt
package com.photovideoeditor.photo.ui

import android.content.Context
import android.graphics.Matrix
import android.graphics.RectF
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.widget.ImageView

/**
 * ImageView supporting pinch-to-zoom and single-finger pan for the photo
 * preview, reporting the current on-screen image rect via [onBoundsChanged]
 * so the crop overlay can stay aligned to it.
 */
class ZoomableImageView(context: Context) : ImageView(context) {
  private val matrixValues = FloatArray(9)
  private var minScale = 1f
  private var maxScale = 4f
  private var lastX = 0f
  private var lastY = 0f
  private var panning = false
  var onBoundsChanged: ((RectF) -> Unit)? = null
  var panZoomEnabled = true

  private val scaleDetector = ScaleGestureDetector(
    context,
    object : ScaleGestureDetector.SimpleOnScaleGestureListener() {
      override fun onScale(detector: ScaleGestureDetector): Boolean {
        val targetScale = (currentScale() * detector.scaleFactor).coerceIn(minScale, maxScale)
        val appliedFactor = targetScale / currentScale()
        imageMatrix = Matrix(imageMatrix).apply { postScale(appliedFactor, appliedFactor, detector.focusX, detector.focusY) }
        onBoundsChanged?.invoke(currentImageBounds())
        return true
      }
    }
  )

  init {
    scaleType = ScaleType.MATRIX
  }

  /** Fits the current drawable to the view bounds and resets zoom/pan. */
  fun resetToFit() {
    val bitmapDrawable = drawable ?: return
    val viewWidth = width.toFloat()
    val viewHeight = height.toFloat()
    val drawableWidth = bitmapDrawable.intrinsicWidth.toFloat()
    val drawableHeight = bitmapDrawable.intrinsicHeight.toFloat()
    if (viewWidth <= 0 || viewHeight <= 0 || drawableWidth <= 0 || drawableHeight <= 0) return
    val scale = minOf(viewWidth / drawableWidth, viewHeight / drawableHeight)
    val dx = (viewWidth - drawableWidth * scale) / 2f
    val dy = (viewHeight - drawableHeight * scale) / 2f
    imageMatrix = Matrix().apply {
      postScale(scale, scale)
      postTranslate(dx, dy)
    }
    minScale = scale
    maxScale = scale * 4f
    onBoundsChanged?.invoke(currentImageBounds())
  }

  override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
    super.onSizeChanged(w, h, oldw, oldh)
    if (w != oldw || h != oldh) resetToFit()
  }

  fun setRenderedBounds(bounds: RectF) {
    val image = drawable ?: return
    imageMatrix = Matrix().apply { setRectToRect(RectF(0f, 0f, image.intrinsicWidth.toFloat(), image.intrinsicHeight.toFloat()), bounds, Matrix.ScaleToFit.FILL) }
  }

  fun currentImageBounds(): RectF {
    val bitmapDrawable = drawable ?: return RectF(0f, 0f, width.toFloat(), height.toFloat())
    val rect = RectF(0f, 0f, bitmapDrawable.intrinsicWidth.toFloat(), bitmapDrawable.intrinsicHeight.toFloat())
    imageMatrix.mapRect(rect)
    return rect
  }

  private fun currentScale(): Float {
    imageMatrix.getValues(matrixValues)
    return matrixValues[Matrix.MSCALE_X]
  }

  override fun onTouchEvent(event: MotionEvent): Boolean {
    if (!panZoomEnabled) return false
    scaleDetector.onTouchEvent(event)
    when (event.actionMasked) {
      MotionEvent.ACTION_DOWN -> {
        lastX = event.x
        lastY = event.y
        panning = true
      }
      MotionEvent.ACTION_POINTER_DOWN -> panning = false
      MotionEvent.ACTION_MOVE -> if (panning && event.pointerCount == 1 && !scaleDetector.isInProgress) {
        val dx = event.x - lastX
        val dy = event.y - lastY
        lastX = event.x
        lastY = event.y
        imageMatrix = Matrix(imageMatrix).apply { postTranslate(dx, dy) }
        onBoundsChanged?.invoke(currentImageBounds())
      }
      MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP -> panning = event.pointerCount <= 1
    }
    return true
  }
}
